staging/src/k8s.io/apiserver/pkg/endpoints/filters/actas.go

```diff
@@ -0,0 +1,172 @@
+package filters
+
+import (
+	"errors"
+	"fmt"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/audit"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/server/httplog"
+	"k8s.io/klog/v2"
+	"net/http"
+	"strings"
+)
+
+const (
+	ActAsUserHeader        = "ActAs-User"
+	ActAsGroupHeader       = "ActAs-Group"
+	ActAsExtraHeaderPrefix = "ActAs-Extra"
+	ActAsUID               = "ActAs-Uid"
+)
+
+func WithActAs(handler http.Handler, a authorizer.Authorizer, s runtime.NegotiatedSerializer) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		actAsUser, isActAsRequest, err := buildActAsUserInfo(req.Header)
+		if err != nil {
+			klog.V(4).Infof("%v", err)
+			responsewriters.InternalError(w, req, err)
+			return
+		}
+		if !isActAsRequest {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		ctx := req.Context()
+		requestor, exists := request.UserFrom(ctx)
+		if !exists {
+			responsewriters.InternalError(w, req, errors.New("no user found for request"))
+			return
+		}
+
+		// check if the real user has actas permission at all
+		actingAsAttributes := &authorizer.AttributesRecord{
+			User:            requestor,
+			Verb:            "actas",
+			Name:            "*",
+			Resource:        "*",
+			ResourceRequest: true,
+		}
+		actAsDecision, reason, err := a.Authorize(ctx, actingAsAttributes)
+		if err != nil || actAsDecision != authorizer.DecisionAllow {
+			klog.V(4).InfoS("Forbidden", "URI", req.RequestURI, "reason", reason, "err", err)
+			responsewriters.Forbidden(ctx, actingAsAttributes, w, req, reason, s)
+			return
+		}
+
+		if actAsUser.Name != user.Anonymous {
+			// When acting as a non-anonymous user, include the 'system:authenticated' group
+			// in the acted user info:
+			// - if no groups were specified
+			// - if a group has been specified other than 'system:authenticated'
+			//
+			// If 'system:unauthenticated' group has been specified we should not include
+			// the 'system:authenticated' group.
+			addAuthenticated := true
+			for _, group := range actAsUser.Groups {
+				if group == user.AllAuthenticated || group == user.AllUnauthenticated {
+					addAuthenticated = false
+					break
+				}
+			}
+
+			if addAuthenticated {
+				actAsUser.Groups = append(actAsUser.Groups, user.AllAuthenticated)
+			}
+		} else {
+			addUnauthenticated := true
+			for _, group := range actAsUser.Groups {
+				if group == user.AllUnauthenticated {
+					addUnauthenticated = false
+					break
+				}
+			}
+
+			if addUnauthenticated {
+				actAsUser.Groups = append(actAsUser.Groups, user.AllUnauthenticated)
+			}
+		}
+
+		// authorize if the acctas user is authorized
+		attribs := authorizer.AttributesRecord{}
+		attribs.User = actAsUser
+		attributes, err := getAuthorizerAttributesFromRequestInfo(ctx, attribs)
```

Comment:
minor: probably use a similar construct to above where we get the normal attributes and replace the user.